graphql/core/validation/rules.py

```diff
@@ -731,8 +737,233 @@ def var_type_allowed_for_type(cls, var_type, expected_type):
+            )
+
+        if not self.same_arguments(ast1.arguments, ast2.arguments):
+            return (
+                (response_name, 'they have differing arguments'),
+                (ast1, ast2)
+            )
+
+        if not self.same_directives(ast1.directives, ast2.directives):
+            return (
+                (response_name, 'they have differing directives'),
+                (ast1, ast2)
+            )
+
+        selection_set1 = ast1.selection_set
+        selection_set2 = ast2.selection_set
+
+        if selection_set1 and selection_set2:
+            visited_fragment_names = set()
+
+            subfield_map = self.collect_field_asts_and_defs(
+                get_named_type(type1),
+                selection_set1,
+                visited_fragment_names
+            )
+
+            subfield_map = self.collect_field_asts_and_defs(
+                get_named_type(type2),
+                selection_set2,
+                visited_fragment_names,
+                subfield_map
+            )
+
+            conflicts = self.find_conflicts(subfield_map)
+            if conflicts:
+                return (
+                    (response_name, [conflict[0] for conflict in conflicts]),
+                    tuple(itertools.chain((ast1, ast2), *[conflict[1] for conflict in conflicts]))
+                )
+
+    def leave_SelectionSet(self, node, *args):
+        field_map = self.collect_field_asts_and_defs(
+            self.context.get_parent_type(),
+            node
+        )
+
+        conflicts = self.find_conflicts(field_map)
+        if conflicts:
+            return [
+                GraphQLError(self.fields_conflict_message(reason_name, reason), list(fields)) for
+                (reason_name, reason), fields in conflicts
+                ]
+
+    @staticmethod
+    def same_type(type1, type2):
+        return type1.is_same_type(type2)
+
+    @staticmethod
+    def same_value(value1, value2):
+        return (not value1 and not value2) or print_ast(value1) == print_ast(value2)
+
+    @classmethod
+    def same_arguments(cls, arguments1, arguments2):
+        # Check to see if they are empty arguments or nones. If they are, we can
+        # bail out early.
+        if not (arguments1 and arguments2):
+            return True
+
+        if len(arguments1) != len(arguments2):
+            return False
+
+        arguments2_values_to_arg = {a.name.value: a for a in arguments2}
+
+        for argument1 in arguments1:
+            argument2 = arguments2_values_to_arg.get(argument1.name.value)
+            if not argument2:
+                return False
+
+            if not cls.same_value(argument1.value, argument2.value):
+                return False
+
+        return True
+
+    @classmethod
+    def same_directives(cls, directives1, directives2):
+        # Check to see if they are empty directives or nones. If they are, we can
+        # bail out early.
+        if not (directives1 and directives2):
```

Comment:
I think this should be `not directives1 and not directives2`. (May add test case for this... One with directive and without directive.)